File: react-native/storyly-placement-react-native/ios/Common/Data/Widgets/VideoFeedData.swift
import Foundation
import StorylyCore
import StorylyVideoFeed

// MARK: - VideoFeed Data Payload Encoding

func encodeVideoFeedDataPayload(_ data: VideoFeedDataPayload) -> [String: Any] {
    return [
        "type": STRWidgetType.videoFeed.description,
        "items": data.items.map { encodeVideoFeedGroup($0) }.compactMap { $0 }
    ]
}

func encodeVideoFeedPayload(_ payload: VideoFeedPayload?) -> [String: Any]? {
    guard let payload = payload else { return nil }
    
    return [
        "component": encodeVideoFeedComponent(payload.component),
        "group": encodeVideoFeedGroup(payload.group),
        "feedItem": encodeVideoFeedItem(payload.feedItem)
    ].compactMapValues { $0 }
}

// MARK: - VideoFeed Group Encoding

func encodeVideoFeedGroup(_ group: VideoFeedGroup?) -> [String: Any]? {
    guard let group = group else { return nil }
    
    return ([
        "id": group.uniqueId,
        "title": group.title,
        "name": group.name,
        "iconUrl": group.iconUrl?.absoluteString,
        "index": group.index,
        "seen": group.seen,
        "type": group.type.description,
        "pinned": group.pinned,
        "nudge": group.nudge,
        "iconVideoUrl": group.iconVideoUrl?.absoluteString,
        "iconVideoThumbnailUrl": group.iconVideoThumbnailUrl?.absoluteString,
        "feedList": group.feedList.map { encodeVideoFeedItem($0) }.compactMap { $0 }
    ] as [String: Any?]).compactMapValues { $0 }
}

// MARK: - VideoFeed Item Encoding

func encodeVideoFeedItem(_ item: VideoFeedItem?) -> [String: Any]? {
    guard let item = item else { return nil }
    
    return ([
        "id": item.uniqueId,
        "title": item.title,
        "name": item.name,
        "index": item.index,
        "seen": item.seen,
        "previewUrl": item.previewUrl?.absoluteString,
        "actionUrl": item.actionUrl,
        "actionProducts": item.actionProducts?.map { encodeSTRProductItem($0) },
        "currentTime": item.currentTime,
        "feedItemComponentList": item.videoFeedItemComponentList
    ] as [String: Any?]).compactMapValues { $0 }
}

// MARK: - VideoFeed Component Encoding

func encodeVideoFeedComponent(_ component: VideoFeedComponent?) -> [String: Any]? {
    guard let component = component else { return nil }
    
    var result: [String: Any] = [
        "id": component.id,
        "type": component.type.get(),
        "customPayload": component.customPayload as Any
    ]
    
    // Add component-specific fields
    switch component {
    case let quiz as VideoFeedQuizComponent:
        result["title"] = quiz.title
        result["options"] = quiz.options
        result["rightAnswerIndex"] = quiz.rightAnswerIndex
        result["selectedOptionIndex"] = quiz.selectedOptionIndex
        
    case let imageQuiz as VideoFeedImageQuizComponent:
        result["title"] = imageQuiz.title
        result["options"] = imageQuiz.options
        result["rightAnswerIndex"] = imageQuiz.rightAnswerIndex
        result["selectedOptionIndex"] = imageQuiz.selectedOptionIndex
        
    case let poll as VideoFeedPollComponent:
        result["title"] = poll.title
        result["options"] = poll.options
        result["selectedOptionIndex"] = poll.selectedOptionIndex
        
    case let emoji as VideoFeedEmojiComponent:
        result["emojiCodes"] = emoji.emojiCodes
        result["selectedEmojiIndex"] = emoji.selectedEmojiIndex
        
    case let rating as VideoFeedRatingComponent:
        result["emojiCode"] = rating.emojiCode
        result["rating"] = rating.rating
        
    case let promoCode as VideoFeedPromoCodeComponent:
        result["text"] = promoCode.text
        
    case let comment as VideoFeedCommentComponent:
        result["text"] = comment.text
        
    case let swipe as VideoFeedSwipeComponent:
        result["text"] = swipe.text
        result["actionUrl"] = swipe.actionUrl
        result["products"] = swipe.products?.map { encodeSTRProductItem($0) }
        
    case let button as VideoFeedButtonComponent:
        result["text"] = button.text
        result["actionUrl"] = button.actionUrl
        result["products"] = button.products?.map { encodeSTRProductItem($0) }
        
    case let productTag as VideoFeedProductTagComponent:
        result["actionUrl"] = productTag.actionUrl
        result["products"] = productTag.products?.map { encodeSTRProductItem($0) }
        
    case let productCard as VideoFeedProductCardComponent:
        result["text"] = productCard.text
        result["actionUrl"] = productCard.actionUrl
        result["products"] = productCard.products?.map { encodeSTRProductItem($0) }
        
    case let productCatalog as VideoFeedProductCatalogComponent:
        result["actionUrlList"] = productCatalog.actionUrlList
        result["products"] = productCatalog.products?.map { encodeSTRProductItem($0) }
    default:
        break
    }
    
    return result.compactMapValues { $0 }
}

File: placement-reactnative/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/widgets/VideoFeedData.kt
package com.storylyplacementreactnative.common.data.widgets

import com.appsamurai.storyly.core.STRWidgetType
import com.appsamurai.storyly.videofeed.data.VideoFeedDataPayload
import com.appsamurai.storyly.videofeed.data.model.VideoFeedPayload
import com.appsamurai.storyly.videofeed.ui.model.*
import com.storylyplacementreactnative.common.data.product.encodeSTRProductItem

fun encodeVideoFeedDataPayload(data: VideoFeedDataPayload): Map<String, Any?> {
    return mapOf(
        "type" to STRWidgetType.VideoFeed.raw,
        "items" to data.items.map { encodeVideoFeedGroup(it) }
    )
}

fun encodeVideoFeedPayload(payload: VideoFeedPayload?): Map<String, Any?>? {
    payload ?: return null
    return mapOf(
        "component" to encodeVideoFeedComponent(payload.component),
        "group" to encodeVideoFeedGroup(payload.group),
        "feedItem" to encodeVideoFeedItem(payload.feedItem),
    )
}

fun encodeVideoFeedGroup(group: VideoFeedGroup?): Map<String, Any?>? {
    group ?: return null
    return mapOf(
        "id" to group.uniqueId,
        "title" to group.title,
        "name" to group.name,
        "iconUrl" to group.iconUrl,
        "index" to group.index,
        "seen" to group.seen,
        "type" to group.type.customName,
        "pinned" to group.pinned,
        "nudge" to group.nudge,
        "iconVideoUrl" to group.iconVideoUrl,
        "iconVideoThumbnailUrl" to group.iconVideoThumbnailUrl,
        "feedList" to group.feedList.map { encodeVideoFeedItem(it) }
    )
}

private fun encodeVideoFeedItem(item: VideoFeedItem?): Map<String, Any?>? {
    item ?: return null
    return mapOf(
        "id" to item.uniqueId,
        "title" to item.title,
        "name" to item.name,
        "index" to item.index,
        "seen" to item.seen,
        "previewUrl" to item.previewUrl,
        "actionUrl" to item.actionUrl,
        "actionProducts" to item.actionProducts?.map { encodeSTRProductItem(it) },
        "currentTime" to item.currentTime,
        "feedItemComponentList" to item.feedItemComponentList
    )
}


fun encodeVideoFeedComponent(component: VideoFeedComponent?): Map<String, Any?>? {
    component ?: return null
    val base = mapOf(
        "id" to component.id,
        "type" to component.type.get(),
        "customPayload" to component.customPayload,
    )

    val additional = when(component) {
        is VideoFeedQuizComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "rightAnswerIndex" to component.rightAnswerIndex,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is VideoFeedImageQuizComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "rightAnswerIndex" to component.rightAnswerIndex,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is VideoFeedPollComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is VideoFeedEmojiComponent -> mapOf(
            "emojiCodes" to component.emojiCodes,
            "selectedEmojiIndex" to component.selectedEmojiIndex,
        )
        is VideoFeedRatingComponent -> mapOf(
            "emojiCode" to component.emojiCode,
            "rating" to component.rating,
        )
        is VideoFeedPromoCodeComponent -> mapOf(
            "text" to component.text,
        )
        is VideoFeedCommentComponent -> mapOf(
            "text" to component.text,
        )
        is VideoFeedSwipeComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is VideoFeedButtonComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is VideoFeedProductTagComponent -> mapOf(
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is VideoFeedProductCardComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is VideoFeedProductCatalogComponent -> mapOf(
            "actionUrlList" to component.actionUrlList,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        else -> emptyMap()
    }

    return base + additional
}